### PR TITLE
add distCp configuration options and force performance improving sett…

### LINKED ISSUE #4377: GC backup and restore: expose distCp configuration options
1. expose distCp [configuration options](https://hadoop.apache.org/docs/r3.2.1/hadoop-distcp/DistCp.html#Command_Line_Options)
2. Use the options to tune distCp performance

### PR DESCRIPTION
**Pull request by talSofer**
Closes #4377

1. expose distCp configuration options 
2. compile gc backup and restore only with the 312 client that uses hadoop 3

### Test results 

With the default property values in this review (See **exceptions** below), GC backup and restore utility copying 800k objects takes ~55 hours to complete with the following breakdown:
1. build file with objects to copy: ~30 min
3. distcp to build file listing: ~16 hours
4. distcp to copy: ~33 hours

### CODE REVIEW

File: clients/spark/core/src/main/scala/io/treeverse/clients/GCBackupAndRestore.scala [hunk 1]
```diff
@@ -78,16 +82,52 @@ object GCBackupAndRestore {
     }
   }
 
+
+
+  // control copy speed
+  //  public static final String CONF_LABEL_DIRECT_WRITE = "distcp.direct.write";
```
Comment (arielshaqed) [inline clients/spark/core/src/main/scala/io/treeverse/clients/GCBackupAndRestore.scala:88]: I'd prefer to have no code in comments.  But if you _really_ want it, could you please expand the comment on line 87?

File: clients/spark/core/src/main/scala/io/treeverse/clients/GCBackupAndRestore.scala [hunk 2]
```diff
@@ -78,16 +82,52 @@ object GCBackupAndRestore {
...
+
   // Construct a DistCp command of the form:
   // `hadoop distcp -D<per storage credentials> -f absoluteAddressesTextFilePath dstNamespaceForHadoopFs`
   // example for command format https://docs.lakefs.io/integrations/distcp.html
+  // with distCp options from https://hadoop.apache.org/docs/r3.2.1/hadoop-distcp/DistCp.html#Command_Line_Options
```
Comment (arielshaqed) [inline clients/spark/core/src/main/scala/io/treeverse/clients/GCBackupAndRestore.scala:93]: GREAT comment!

File: clients/spark/core/src/main/scala/io/treeverse/clients/GCBackupAndRestore.scala [hunk 3]
```diff
@@ -78,16 +82,52 @@ object GCBackupAndRestore {
...
-      dstNamespaceForHadoopFs: String
-  ): Array[String] = {
+                              hadoopProps: Array[(String, String)],
+                              absoluteAddressesTextFilePath: String,
+                              dstNamespaceForHadoopFs: String,
+                              hc: Configuration,
+                              numObjectsToCopy: Long
+                            ): Array[String] = {
+
+    val distCpLogsPath = hc.get(DistCpConstants.CONF_LABEL_LOG_PATH, dstNamespaceForHadoopFs + "_distCp/logs/")
```
Comment (arielshaqed) [inline clients/spark/core/src/main/scala/io/treeverse/clients/GCBackupAndRestore.scala:102]: This assumes a trailing slash on `dstNamespaceForHadoopFs`.  Perhaps we can/should use a class such as Path or URI and resolve or join components?

File: clients/spark/core/src/main/scala/io/treeverse/clients/GCBackupAndRestore.scala [hunk 4]
```diff
@@ -78,16 +82,52 @@ object GCBackupAndRestore {
...
+    print("distCpLogsPath: " + distCpLogsPath + "\n")
```
Comment (arielshaqed) [inline clients/spark/core/src/main/scala/io/treeverse/clients/GCBackupAndRestore.scala:103]:
```suggestion
    println("distCpLogsPath: " + distCpLogsPath)
```
or
```suggestion
    println(s"distCpLogsPath ${distCpLogsPath}")
```
(and similarly below).

Also consider first fetching all the numbers, then  outputting them in one go.

File: clients/spark/core/src/main/scala/io/treeverse/clients/GCBackupAndRestore.scala [hunk 5]
```diff
@@ -78,16 +82,52 @@ object GCBackupAndRestore {
...
+    print("numListstatusThreads: " + numListstatusThreads + "\n")
+
+    // Tune distCp options that control the speed of the copy stage
+    // https://docs.cloudera.com/HDPDocuments/HDP3/HDP-3.1.4/bk_cloud-data-access/content/distcp-perf-mappers.html
+    val maxMaps = hc.get(DistCpConstants.CONF_LABEL_MAX_MAPS, max(DistCpConstants.DEFAULT_MAPS, numObjectsToCopy / 1000).toString) //TODO: what is a reasonable number of maps
+    print("maxMaps: " + maxMaps + "\n")
+    val mapsBandwidth = hc.get(DistCpConstants.CONF_LABEL_BANDWIDTH_MB, DistCpConstants.DEFAULT_BANDWIDTH_MB.toString)
+    print("mapsBandwidth: " + mapsBandwidth + "\n")
+
     hadoopProps.map((prop) => "-D" + prop._1 + "=" + prop._2) ++
```
Comment (arielshaqed) [inline clients/spark/core/src/main/scala/io/treeverse/clients/GCBackupAndRestore.scala:116]: Consider
```suggestion
    hadoopProps.map((prop) => s"-D${prop._1}=${prop._2}") ++
```
or something like
```suggestion
    hadoopProps.map({ case (k, v) => s"-D$k=$v" }) ++
```

File: clients/spark/core/src/main/scala/io/treeverse/clients/GCBackupAndRestore.scala [hunk 6]
```diff
@@ -138,9 +179,9 @@ object GCBackupAndRestore {
 
   /** Required arguments are the following:
    *  1. address of parquet that includes the relative paths of files to backup or restore, created by a gc run
-   *  2. src: the namespace to backup or restore objects from/to (i.e. repo storage namespace, or an external location compatibly)
-   *  3. backup/restore destination: the namespace to backup or restore objects from/to (i.e. an external location or repo storage namespace compatibly)
-   *  4. Object storage type: "s3" or "azure"
+   *     2. src: the namespace to backup or restore objects from/to (i.e. repo storage namespace, or an external location compatibly)
```
Comment (arielshaqed) [inline clients/spark/core/src/main/scala/io/treeverse/clients/GCBackupAndRestore.scala:182]: I _think_ that this changes indentation inside the comment halfway through the list ("1." seems unindented)...

File: clients/spark/core/src/main/scala/io/treeverse/clients/GCBackupAndRestore.scala [hunk 7]
```diff
@@ -78,16 +82,52 @@ object GCBackupAndRestore {
...
+    val distCpLogsPath = hc.get(DistCpConstants.CONF_LABEL_LOG_PATH, dstNamespaceForHadoopFs + "_distCp/logs/")
+    print("distCpLogsPath: " + distCpLogsPath + "\n")
+
+    // Tune distCp options that control the speed of the file-to-copy list building stage
+    val numListstatusThreads = hc.get(DistCpConstants.CONF_LABEL_LISTSTATUS_THREADS, DistCpMaxNumListStatusThreads.toString)
+    print("numListstatusThreads: " + numListstatusThreads + "\n")
+
+    // Tune distCp options that control the speed of the copy stage
+    // https://docs.cloudera.com/HDPDocuments/HDP3/HDP-3.1.4/bk_cloud-data-access/content/distcp-perf-mappers.html
+    val maxMaps = hc.get(DistCpConstants.CONF_LABEL_MAX_MAPS, max(DistCpConstants.DEFAULT_MAPS, numObjectsToCopy / 1000).toString) //TODO: what is a reasonable number of maps
```
Comment (arielshaqed) [inline clients/spark/core/src/main/scala/io/treeverse/clients/GCBackupAndRestore.scala:111]: I don't understand: why do we want 1 copy thread per object?

**Review by arielshaqed — CHANGES_REQUESTED** [review]
Requested for comment or change:

1. What are the resulting execution times?
1. I don't understand why the code seems to say 1 thread per object.
It might just be a confusing variable name.
1. We might have a trailing-slash-on-namespace issue, see comment

Other minor cleanups are of course not blocking.

File: clients/spark/core/src/main/hadoop3/scala/io/treeverse/clients/conditional/GCBackupAndRestore.scala [hunk 8]
```diff
@@ -81,13 +86,53 @@ object GCBackupAndRestore {
...
+        .translateURI(URI.create(distCpLogsPath), storageType)
+        .normalize()
+        .toString
+    // Tune distCp options that control the speed of the file-to-copy list building stage
+    val numListstatusThreads =
+      hc.get(DistCpConstants.CONF_LABEL_LISTSTATUS_THREADS, DistCpMaxNumListStatusThreads.toString)
+    // Tune distCp options that control the speed of the copy stage
+    // https://docs.cloudera.com/HDPDocuments/HDP3/HDP-3.1.4/bk_cloud-data-access/content/distcp-perf-mappers.html
+    val maxMaps = hc.get(DistCpConstants.CONF_LABEL_MAX_MAPS,
+                         max(DistCpConstants.DEFAULT_MAPS, numObjectsToCopy / 1000).toString
```
Comment (arielshaqed) [inline clients/spark/core/src/main/hadoop3/scala/io/treeverse/clients/conditional/GCBackupAndRestore.scala:113]: Please round up the division.  While we are protected from the worst cases (getting 0...) by DEFAULT_MAPS, if we round up then we know we get up to 1000 objects per map.  Right now we're rounding down, so we know we get at least 1000 objects per map but it's hard to figure out exactly how many.

File: clients/spark/core/src/main/hadoop3/scala/io/treeverse/clients/conditional/GCBackupAndRestore.scala [hunk 9]
```diff
@@ -138,9 +184,9 @@ object GCBackupAndRestore {
 
   /** Required arguments are the following:
    *  1. address of parquet that includes the relative paths of files to backup or restore, created by a gc run
-   *  2. src: the namespace to backup or restore objects from/to (i.e. repo storage namespace, or an external location compatibly)
-   *  3. backup/restore destination: the namespace to backup or restore objects from/to (i.e. an external location or repo storage namespace compatibly)
-   *  4. Object storage type: "s3" or "azure"
+   *     2. src: the namespace to backup or restore objects from/to (i.e. repo storage namespace, or an external location compatibly)
+   *     3. backup/restore destination: the namespace to backup or restore objects from/to (i.e. an external location or repo storage namespace compatibly)
+   *     4. Object storage type: "s3" or "azure"
```
Comment (arielshaqed) [inline clients/spark/core/src/main/hadoop3/scala/io/treeverse/clients/conditional/GCBackupAndRestore.scala:189]: I think this still clobbers indentation compared to "1" above.

**Review by arielshaqed — APPROVED** [review]
Thanks!

This is a really nice wrapper for distCP that seems much easier to use, and is happily approved.  But please do go over comments before pulling. :-)

File: clients/spark/core/src/main/hadoop3/scala/io/treeverse/clients/conditional/GCBackupAndRestore.scala (hunk 8 above)
Comment (talSofer) [reply clients/spark/core/src/main/hadoop3/scala/io/treeverse/clients/conditional/GCBackupAndRestore.scala]: thx, I decided to use the default num of max maps given by distCp so this line changed

File: clients/spark/core/src/main/hadoop3/scala/io/treeverse/clients/conditional/GCBackupAndRestore.scala [hunk 10]
```diff
@@ -81,13 +86,53 @@ object GCBackupAndRestore {
...
+        .normalize()
+        .toString
+    // Tune distCp options that control the speed of the file-to-copy list building stage
+    val numListstatusThreads =
+      hc.get(DistCpConstants.CONF_LABEL_LISTSTATUS_THREADS, DistCpMaxNumListStatusThreads.toString)
+    // Tune distCp options that control the speed of the copy stage
+    // https://docs.cloudera.com/HDPDocuments/HDP3/HDP-3.1.4/bk_cloud-data-access/content/distcp-perf-mappers.html
+    val maxMaps = hc.get(DistCpConstants.CONF_LABEL_MAX_MAPS,
+                         max(DistCpConstants.DEFAULT_MAPS, numObjectsToCopy / 1000).toString
+                        ) //TODO: what is a reasonable number of maps
```
[the remark this answers is not in the document]
Comment (talSofer) [reply clients/spark/core/src/main/hadoop3/scala/io/treeverse/clients/conditional/GCBackupAndRestore.scala]: thx, I decided to use the default num of max maps given by distCp so this line changed

File: clients/spark/core/src/main/hadoop3/scala/io/treeverse/clients/conditional/GCBackupAndRestore.scala (hunk 9 above)
Comment (talSofer) [reply clients/spark/core/src/main/hadoop3/scala/io/treeverse/clients/conditional/GCBackupAndRestore.scala]: fixed

**Pull request merged**